design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,474 @@
+
+Note that just because the customer added a label for us to filter over doesn't mean that all secret entries in the K8s Secret are stored in Conjur but it gives us a way to perform a more focused search for "conjur-map" in the K8s Secrets.
+
+Secrets Provider Manifest with the `K8S_SECRET_LABEL` environment variable:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: secrets-provider-deployment
+spec:
+  replicas: 1
+  template:
+    spec:
+      serviceAccountName: sa-1
+      containers:
+      - image: secrets-provider:latest
+        name: cyberark-secrets-provider
+   ...
+      env:
+        - name: K8S_SECRET_LABELS
+          value: 'configurable-key:env-prod,environment:prod'
+```
+
+These labels will have an OR inclusive relationship, not an AND relationship. A customer can also write '*' to imply fetching all K8s secrets in the namespace. If K8S_SECRET_LABELS is not supplied, by default, the Secrets Provider will fetch all K8s Secrets in the namespace.
+
+To read the labels set on K8s Secrets, the Service Account used for the Secrets Provider will need "list" privileges on K8s Secrets resources.
+
+The Service Account / Role / RoleBinding will be defined as the following:
+
+```yaml
+# Service Account definition 
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: conjur-sa-1
+
+# Role definition
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  namespace: app-namespace
+  name: secrets-access
+rules:
+- apiGroups: [""]
+  resources: ["secrets"]
+  verbs: ["get", "update", "list"] # new privilege here
+
+# RoleBinding definition to associate the SA with the Role
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: secrets-access-binding
+  namespace: app-namespace
+subjects:
+  - kind: ServiceAccount
+    name: conjur-sa-1
+    apiGroup: rbac.authorization.k8s.io
+roleRef:
+  kind: Role
+  name: secrets-access
+  apiGroup: rbac.authorization.k8s.io
+```
+
+#### Code changes
+
+Although we will now offer "K8S_SECRET_LABELS" for easy filtering on K8s Secrets, we will still support the "K8S_SECRET" environment variable if the customer prefers to list the K8s Secrets they need values for from Conjur. 
+
+### Enhancements
+
+As optional (yet important) enhancements, we suggest the following:
+
+1. #### Differentiate between deployment types using K8s API
+
+   **Current solution**: Customer supplies this value in `CONTAINER_MODE` env var to Secrets Provider.
+
+   **Enhanced solution**: Use K8s API `GET namespaces/{namespace}/pods/{pod}` to get the pod's manifest and derive the deployment used for Secrets Provider.
+
+   **Motivation**: better UX, no need to maintain another env var. Also, ensures correct value and prevent mistakes.
+   **Requirements**: Add `get` rights for `pods` to Secrets Provider's service account in K8s.
```

Comment:
when you create this issue, please address backwards compatibilty. How do we handle a case where the service-account doesn't have this privilege?

Reply:
you are right! Added the following
```
K8S_SECRETS exists in manifest?
  YES? Fetch k8s secrets defined under K8S_SECRETS
  NO? K8S_SECRETS_LABEL exists?
    YES? Fetch k8s secrets with label defined under K8S_SECRETS_LABEL
    NO? Log failure
```

Reply:
Log failure and exit